sumpy/tools.py

```diff
@@ -84,28 +86,112 @@ def mi_power(vector, mi, evaluate=True):
+        return expr
+
+    name = sac.assign_unique("temp", expr)
+    return sym.Symbol(name)
+
+
+class MiDerivativeTaker(object):
+
+    def __init__(self, expr, var_list, rscale=1, sac=None):
+        r"""
+        A class to take scaled derivatives of the symbolic expression
+        expr w.r.t. variables var_list and the scaling parameter rscale.
+
+        Consider a Taylor multipole expansion:
+
+        .. math::
+
+            f (x - y) = \sum_{i = 0}^{\infty} (\partial_y^i f) (x - y) \big|_{y = c}
+           \frac{(y - c)^i}{i!} .
+
+        Now suppose we would like to use a scaled version :math:`g` of the
+        kernel :math:`f`:
+
+        .. math::
+
+            \begin{eqnarray*}
+              f (x) & = & g (x / \alpha),\\
+              f^{(i)} (x) & = & \frac{1}{\alpha^i} g^{(i)} (x / \alpha) .
+            \end{eqnarray*}
+
+        where :math:`\alpha` is chosen to be on a length scale similar to
+        :math:`x` (for example by choosing :math:`\alpha` proporitional to the
+        size of the box for which the expansion is intended) so that :math:`x /
+        \alpha` is roughly of unit magnitude, to avoid arithmetic issues with
+        small arguments. This yields
+
+        .. math::
+
+            f (x - y) = \sum_{i = 0}^{\infty} (\partial_y^i g)
+            \left( \frac{x - y}{\alpha} \right) \Bigg|_{y = c}
+            \cdot
+            \frac{(y - c)^i}{\alpha^i \cdot i!}.
+
+        Observe that the :math:`(y - c)` term is now scaled to unit magnitude,
+        as is the argument of :math:`g`.
+
+        With :math:`\xi = x / \alpha`, we find
+
+        .. math::
+
+            \begin{eqnarray*}
+              g (\xi) & = & f (\alpha \xi),\\
+              g^{(i)} (\xi) & = & \alpha^i f^{(i)} (\alpha \xi) .
+            \end{eqnarray*}
+
+        Generically for all kernels, :math:`f^{(i)} (\alpha \xi)` is computable
+        by taking a sufficient number of symbolic derivatives of :math:`f` and
+        providing :math:`\alpha \xi = x` as the argument.
+
+        Now, for some kernels, like :math:`f (x) = C \log x`, the powers of
+        :math:`\alpha^i` from the chain rule cancel with the ones from the
+        argument substituted into the kernel derivatives:
+
+        .. math::
+
+            g^{(i)} (\xi) = \alpha^i f^{(i)} (\alpha \xi) = C' \cdot \alpha^i \cdot
+            \frac{1}{(\alpha x)^i} \quad (i > 0),
+
+        making them what you might call *scale-invariant*.
+
+        This derivative taker returns :math:`g^{(i)}(\xi) = \alpha^i f^{(i)}`
+        given :math:`f^{(0)}` as *expr* and :math:`\alpha` as *rscale*.
+        """
 
-    def __init__(self, expr, var_list):
         assert isinstance(expr, sym.Basic)
         self.var_list = var_list
         empty_mi = (0,) * len(var_list)
         self.cache_by_mi = {empty_mi: expr}
+        self.rscale = rscale
+        self.sac = sac
+        self.dim = len(self.var_list)
+        self.orig_expr = expr
```

Comment:
Maybe underscore these based on which are supposed to be exposed and not?

Reply:
I guess none of them? Only `.diff` method is supposed to be used by the user.